Index: server/rendering/UserCamera.cc
===================================================================
--- server/rendering/UserCamera.cc	(revision 8861)
+++ server/rendering/UserCamera.cc	(working copy)
@@ -78,8 +78,8 @@
 {
   OgreCamera::LoadCam(node);
 
-  this->SetFOV( DTOR(60) );
-  this->SetClipDist(0.01, 50);
+  this->SetFOV( DTOR(90) );
+  this->SetClipDist(0.001, 1000);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
