Index: server/Plugin.hh
===================================================================
--- server/Plugin.hh	(revision 8824)
+++ server/Plugin.hh	(working copy)
@@ -28,8 +28,8 @@
 }
 
 #define GZ_REGISTER_PLUGIN(name, classname) \
-extern "C" Plugin *RegisterPlugin(); \
-Plugin *RegisterPlugin() \
+extern "C" gazebo::Plugin *RegisterPlugin(); \
+gazebo::Plugin *RegisterPlugin() \
 {\
   return new classname();\
 }
