Index: cmake/SearchForStuff.cmake
===================================================================
--- cmake/SearchForStuff.cmake	(revision 8861)
+++ cmake/SearchForStuff.cmake	(working copy)
@@ -1,9 +1,6 @@
 include (${gazebo_cmake_dir}/GazeboUtils.cmake)
 include (CheckCXXSourceCompiles)
 
-# John - ode version with joint damping
-SET(ODE_JOINT_DAMPING_VERSION 0.11.1.1 CACHE INTERNAL "ODE version with joint damping" FORCE)
-
 set (INCLUDE_WEBGAZEBO ON CACHE BOOL "Build webgazebo" FORCE)
 set (OGRE_LIBRARY_PATH "/usr/local/lib" CACHE INTERNAL "Ogre library path")
 
@@ -73,6 +70,15 @@
                           ${ODE_LDFLAGS})
   ENDIF (NOT ODE_FOUND)
 
+  # patched ode version with joint damping
+  SET(ODE_JOINT_DAMPING_VERSION 0.11.1.68 CACHE INTERNAL "ODE version with joint damping" FORCE)
+  pkg_check_modules(ODE_JOINT_DAMPING ode>=${ODE_JOINT_DAMPING_VERSION})
+  IF (NOT ODE_JOINT_DAMPING_FOUND)
+    SET (INCLUDE_ODE_JOINT_DAMPING FALSE CACHE BOOL "No support for ODE damping")
+  ELSE (NOT ODE_JOINT_DAMPING_FOUND)
+    SET (INCLUDE_ODE_JOINT_DAMPING TRUE CACHE BOOL "Include support for ODE damping")
+  ENDIF (NOT ODE_JOINT_DAMPING_FOUND)
+
   #################################################
   # Find OGRE 
   pkg_check_modules(OGRE-RTShaderSystem OGRE-RTShaderSystem>=${MIN_OGRE_VERSION})
Index: config.h.in
===================================================================
--- config.h.in	(revision 8861)
+++ config.h.in	(working copy)
@@ -14,3 +14,4 @@
 #cmakedefine ENABLE_SHADOWS 1
 #cmakedefine INCLUDE_BULLET 1
 #cmakedefine INCLUDE_ODE 1
+#cmakedefine INCLUDE_ODE_JOINT_DAMPING 1
